src/shared/bpf-program.c

```diff
@@ -41,6 +42,56 @@ DEFINE_STRING_TABLE_LOOKUP(bpf_cgroup_attach_type, int);
 
 DEFINE_HASH_OPS_WITH_KEY_DESTRUCTOR(bpf_program_hash_ops, void, trivial_hash_func, trivial_compare_func, bpf_program_free);
 
+int bpf_program_supported(void) {
+        static bool cached = false;
+        static int result = 0;
+
+        /* Currently, we only use the following three types:
+        * - BPF_PROG_TYPE_CGROUP_SKB, supported since kernel v4.10 (0e33661de493db325435d565a4a722120ae4cbf3),
+        * - BPF_PROG_TYPE_CGROUP_DEVICE, supported since kernel v4.15 (ebc614f687369f9df99828572b1d85a7c2de3d92),
+        * - BPF_PROG_TYPE_CGROUP_SOCK_ADDR, supported since kernel v4.17 (4fbac77d2d092b475dda9eea66da674369665427).
+        * Hence, as our baseline on the kernel is v5.4, it is enough to check only if CONFIG_CGROUP_BPF is
+        * ebabled on the kernel (see comments below). */
+
+        if (cached)
+                return result;
+
+        cached = true;
+
+        /* Unfortunately the kernel allows us to create BPF_PROG_TYPE_CGROUP_SKB (and other types, maybe?)
+         * programs even when CONFIG_CGROUP_BPF is turned off at kernel compilation time. This sucks o
+         * f course: why does it allow us to create a cgroup BPF program if we can't do a thing with it later?
+         *
+         * We detect this case by issuing the BPF_PROG_DETACH bpf() call with invalid file descriptors: if
+         * CONFIG_CGROUP_BPF is turned off, then the call will fail early with EINVAL. If it is turned on the
+         * parameters are validated however, and that'll fail with EBADF then.
+         *
+         * The check seems also important when we are running with sanitizers. With sanitizers (at least with
+         * LLVM v20), the following check and other bpf() calls fails even if the kernel supports BPF. To
+         * avoid unexpected fail when running with sanitizers, let's explicitly check if bpf() syscall works. */
+
+        // FIXME: Clang doesn't 0-pad with structured initialization, causing
```

Comment:
Not your fault, but this is outdated. Since gcc 15 the default behavior has been changed to not zero'ify padding (c f. https://gcc.gnu.org/gcc-15/changes.html)

Please drop it hence